spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/procedures/GenerateChangesProcedure.java

```diff
@@ -0,0 +1,210 @@
+    String tableName = args.getString(0);
+
+    // Read data from the table.changes
+    Dataset<Row> df = changelogRecords(tableName, args);
+
+    // Compute the pre-image and post-images if the identifier columns are provided.
+    if (!args.isNullAt(4)) {
+      String[] identifierColumns = args.getString(4).split(",");
+      if (identifierColumns == null || identifierColumns.length > 0) {
+        df = withUpdate(df, identifierColumns);
+      }
+    }
+
+    String viewName = viewName(args, tableName);
+
+    // Create a view for users to query
+    df.createOrReplaceTempView(viewName);
+
+    return toOutputRows(viewName);
+  }
+
+  private Dataset<Row> changelogRecords(String tableName, InternalRow args) {
+    DataFrameReader reader = spark().read();
+
+    // we don't have to validate the snapshot ids here because the reader will do it for us.
+    if (!args.isNullAt(1)) {
+      long startSnapshotId = args.getLong(1);
+      reader = reader.option(SparkReadOptions.START_SNAPSHOT_ID, startSnapshotId);
+    }
+
+    if (!args.isNullAt(2)) {
+      long endSnapshotId = args.getLong(2);
+      reader = reader.option(SparkReadOptions.END_SNAPSHOT_ID, endSnapshotId);
+    }
+
+    return reader.table(tableName + "." + SparkChangelogTable.TABLE_NAME);
+  }
+
+  @NotNull
+  private static String viewName(InternalRow args, String tableName) {
+    String viewName = args.isNullAt(3) ? null : args.getString(3);
+    if (viewName == null) {
+      String shortTableName =
+          tableName.contains(".") ? tableName.substring(tableName.lastIndexOf(".") + 1) : tableName;
+      viewName = shortTableName + "_changes";
+    }
+    return viewName;
+  }
+
+  private Dataset<Row> withUpdate(Dataset<Row> df, String[] identifiers) {
+    Column[] partitionSpec = getPartitionSpec(df, identifiers);
+
+    Dataset<Row> dfWithUpdate =
+        df.withColumn("count", functions.count("*").over(Window.partitionBy(partitionSpec)))
+            .withColumn(
+                "rank",
+                functions
+                    .rank()
+                    .over(
+                        Window.partitionBy(partitionSpec)
+                            .orderBy(MetadataColumns.CHANGE_TYPE.name())));
+
+    Dataset<Row> preImageDf =
+        dfWithUpdate
+            .filter("rank = 1")
+            .filter("count = 2")
+            .drop("rank", "count")
+            .withColumn(
+                MetadataColumns.CHANGE_TYPE.name(),
+                functions.lit(ChangelogOperation.UPDATE_PREIMAGE.name()));
+
+    Dataset<Row> postImageDf =
+        dfWithUpdate
+            .filter("rank = 2")
+            .filter("count = 2")
+            .drop("rank", "count")
+            .withColumn(
+                MetadataColumns.CHANGE_TYPE.name(),
+                functions.lit(ChangelogOperation.UPDATE_POSTIMAGE.name()));
+
+    // remove the carry-over rows
+    Dataset<Row> dfWithoutCarryOver = removeCarryOvers(preImageDf.union(postImageDf));
```

Comment:
Should we make it optional since it is a heavy operation?

Reply:
Is there another algorithm we can consider that would make it cheaper? Will something like this work?

```
- Load DELETEs and INSERTs as a DF
- Repartition the DF by primary key, _change_ordinal and locally sort by primary key, _change_ordinal, _operation_type
- Call mapPartitions with a closure that would look at the previous, current and next rows
  - If the previous, current, next row keys are different, output the current row as-is
  - If the next row key is same, the current row must be DELETE and the next row must be INSERT (if not -> exception)
      - If other columns beyond the key are same, it is a copied over row
          - Output null if unchanged rows should be ignored
          - Output the current row as-is if all rows should be produced 
      - If other columns beyond key are different, it is an update
          - Output the current row as pre-update
  - If the previous row key is same as the current one, the current row must be INSERT and the previous row must be DELETE
      - If other columns beyond the key are same, it is a copied over row
          - Output null if unchanged rows should be ignored
          - Output the current row as-is if all rows should be produced 
      - If other columns beyond key are different, it is an update
          - Output the current row as post-update
```

That would require reading the changes only once, doing a single hash-based shuffle to co-locate rows for the same key and change ordinal, keeping  at most 3 rows in memory at a time. Seems fairly cheap?

Reply:
Thanks @aokolnychyi for the suggestion. Make sense to shuffle once. Agreed with @rdblue, to just look forward should be good, no need to search bidirectionally. Will make the change accordingly.

Reply:
Made the change, could you take a look?

Reply:
I agree with can pre-compute the required state by just looking back.